### PR TITLE
Dimension Reduce calculator

### PR DESCRIPTION
**Pull request by martin-hora**

### CODE REVIEW

File: boltview/reduce.tcc [hunk 1]
```diff
@@ -296,27 +325,32 @@ struct ReduceImplementation {
...
+		dim3 grid(block_count, reduced_size[0], reduced_size.kDimension >= 2 ? reduced_size[1] : 1);
+		
+		if (block_count > 1) {
+			dimensionReduceKernel<TView, decltype(current_tmp), TOutputValue, TOperator, tDimension, kBlockSize><<<grid, block, 0, execution_policy.cuda_stream>>>(inview, current_tmp, initial_value, reduction_operator);
 			BOLT_CHECK_ERROR_AFTER_KERNEL("dimensionReduceKernel tmp buffer", grid, block);
-			run(tmp_buffer.constView(), output_view, dimension, initial_value, reduction_operator, execution_policy);
+			run(current_tmp, following_tmp, output_view, dimension, initial_value, reduction_operator, execution_policy);
 		} else {
-			dim3 block(kBlockSize, 1, 1);
-			dim3 grid(1 + (size[tDimension] - 1) / block.x, reduced_size[0], reduced_size.kDimension >= 2 ? reduced_size[1] : 1);
```
Comment (martin-hora) [inline boltview/reduce.tcc:313]: Why was the x-coordinate of grid size calculated differently in this case?

File: boltview/reduce.tcc [hunk 2]
```diff
@@ -405,6 +439,85 @@ public:
...
+		execution_policy_(execution_policy),
+		tmp_image_(detail::estimateTemporaryBuffer(size, tReduceDimension, kBlockSize, kBucketSize))
+	{}
+
+	template<typename TInView, typename TOutView, typename TOperator>
+	void runAsync(
+		TInView input, TOutView output,
+		TOutputValue initial_value, TOperator reduction_operator
+	) {
+		static_assert(TInView::kDimension == tInputDimension, "Invalid input dimension.");
```
Comment (martin-hora) [inline boltview/reduce.tcc:497]: We should probably check that the input.size() <= size from the constructor.

File: boltview/reduce.tcc (hunk 1 above)
Comment (JanKolomaznik) [reply boltview/reduce.tcc:313]: seems like a bug - it should be 1 (or block_count)

File: boltview/reduce.tcc (hunk 2 above)
Comment (JanKolomaznik) [reply boltview/reduce.tcc:497]: sure - throw an exception (or at least assert)

**Pull request merged**